usr/lib/common/key_mgr.c

```diff
@@ -42,7 +42,65 @@
 
 #include <openssl/crypto.h>
 
-static CK_BBOOL true = TRUE, false = FALSE;
+static CK_BBOOL true = TRUE;
+
+CK_RV key_mgr_apply_always_sensitive_never_extractable_attrs(
+                                    STDLL_TokData_t *tokdata, OBJECT *key_obj)
+{
+    CK_BBOOL ck_true = TRUE;
+    CK_ATTRIBUTE *new_attr = NULL;
+    CK_BBOOL flag;
+    CK_RV rc;
+
+    UNUSED(tokdata);
+
+    rc = template_attribute_get_bool(key_obj->template, CKA_SENSITIVE, &flag);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to find CKA_SENSITIVE in key object template.\n");
+        goto error;
+    }
+
+    rc = build_attribute(CKA_ALWAYS_SENSITIVE, &flag, sizeof(CK_BBOOL),
+                         &new_attr);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build attribute failed.\n");
+        goto error;
+    }
+    rc = template_update_attribute(key_obj->template, new_attr);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("template_update_attribute failed.\n");
+        goto error;
+    }
+    new_attr = NULL;
+
+    rc = template_attribute_get_bool(key_obj->template, CKA_EXTRACTABLE, &flag);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to find CKA_EXTRACTABLE in key object template.\n");
+        goto error;
+    }
+
+    rc = build_attribute(CKA_NEVER_EXTRACTABLE, &ck_true, sizeof(CK_BBOOL),
+                         &new_attr);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
+        goto error;
+    }
+    if (flag == TRUE)
+        *(CK_BBOOL *)new_attr->pValue = FALSE;
+
+    rc = template_update_attribute(key_obj->template, new_attr);
```

Comment:
Right, the callers (I checked all 4) will free the entire object (including the template) on error, so an inconsistent template does not matter in case of an error.